src/swap/changehero.js

```diff
@@ -0,0 +1,412 @@
+  wallet: EdgeCurrencyWallet,
+  currencyCode: string
+): Promise<string> {
+  const addressInfo = await wallet.getReceiveAddress({ currencyCode })
+  return addressInfo.legacyAddress && !dontUseLegacy[currencyCode]
+    ? addressInfo.legacyAddress
+    : addressInfo.publicAddress
+}
+
+function checkReply(reply: Object, request: EdgeSwapRequest) {
+  if (reply.error != null) {
+    if (
+      reply.error.code === -32602 ||
+      /Invalid currency:/.test(reply.error.message)
+    ) {
+      throw new SwapCurrencyError(
+        swapInfo,
+        request.fromCurrencyCode,
+        request.toCurrencyCode
+      )
+    }
+    throw new Error('ChangeHero error: ' + JSON.stringify(reply.error))
+  }
+}
+
+export function makeChangeHeroPlugin(
+  opts: EdgeCorePluginOptions
+): EdgeSwapPlugin {
+  const { initOptions, io, log } = opts
+  const { fetchCors = io.fetch } = io
+
+  if (initOptions.apiKey == null) {
+    throw new Error('No ChangeHero apiKey or secret provided.')
+  }
+  const { apiKey } = initOptions
+
+  async function call(json: any) {
+    const body = JSON.stringify(json)
+
+    const headers = {
+      'Content-Type': 'application/json',
+      'api-key': apiKey
+    }
+    const response = await fetchCors(uri, { method: 'POST', body, headers })
+
+    if (!response.ok) {
+      throw new Error(`ChangeHero returned error code ${response.status}`)
+    }
+    return response.json()
+  }
+
+  const out: EdgeSwapPlugin = {
+    swapInfo,
+    async fetchSwapQuote(
+      request: EdgeSwapRequest,
+      userSettings: Object | void
+    ): Promise<EdgeSwapQuote> {
+      const fixedPromise = this.getFixedQuote(request, userSettings)
+      // Estimated swaps disabled
+      const fixedResult = await fixedPromise
+      return fixedResult
+    },
+    async getFixedQuote(
+      request: EdgeSwapRequest,
+      userSettings: Object | void
+    ): Promise<EdgeSwapQuote> {
+      const [fromAddress, toAddress] = await Promise.all([
+        getAddress(request.fromWallet, request.fromCurrencyCode),
+        getAddress(request.toWallet, request.toCurrencyCode)
+      ])
+      const quoteAmount =
+        request.quoteFor === 'from'
+          ? await request.fromWallet.nativeToDenomination(
+              request.nativeAmount,
+              request.fromCurrencyCode
+            )
+          : await request.toWallet.nativeToDenomination(
+              request.nativeAmount,
+              request.toCurrencyCode
+            )
+
+      const { safeFromCurrencyCode, safeToCurrencyCode } = safeCurrencyCodes(
+        CURRENCY_CODE_TRANSCRIPTION,
+        request
+      )
+
+      const fixedRateQuote = await call({
+        jsonrpc: '2.0',
+        id: 'one',
+        method: 'getFixRate',
+        params: {
+          from: safeFromCurrencyCode,
+          to: safeToCurrencyCode
+        }
+      })
+
+      const [{ id: responseId }] = fixedRateQuote.result
+      const params =
+        request.quoteFor === 'from'
```

Comment:
API does not seem to be able to specify the network chain separately from the token code. This is mandatory for integration. See the LetsExchange plugin here

https://github.com/EdgeApp/edge-exchange-plugins/blob/master/src/swap/letsexchange.js#L160

Reply:
Do you mean that we need to add parameters network_from and network_to to our api support?

Reply:
we specify network in tickers usually, that's our current idea. Also I see that another exchangers use INVALID_CURRENCY_CODES, for instance here https://github.com/EdgeApp/edge-exchange-plugins/blob/master/src/swap/changelly.js#L28 and no network in changelly plugin.

Reply:
Yes Changelly doesn’t have network support but we are not allowing new plugins to not have an explicit network field. As well we’ll be deprecating any plug-ins that don’t have that feature. It causes too many support issues and apparent loss of funds.

Reply:
But if you need to exchange for instance USDC on MATIC mainnet you have to use not USDC ticker, but USDCPOLYGON. That's our idea. Is there any way to specify some logic of such conversion from USDC to USDCPOLYGON in exchange plugin, but not on ChangeHero API side?

Reply:
While that’s doable, it’s a hack and doesn’t scale to support a lot of tokens. As well we would have to update the app every time you add new tokens. 

There can also be errors by which a new chain also supports USDC but you haven’t added a ticker conversion for it and the user then receives ETH USDC.

Reply:
That's tricky for us to implement at the current times. Is there any way to disable all chains except ETH to proceed with the release while we are finding a solution? Is there any way to deliver such functionality in the future Edge wallet releases?

Reply:
Sure. if you compare the `fromWallet.currencyInfo.currencyCode` to the `fromCurrencyCode` (and do the same for the `to` values, they should be the same and can only be different if `fromWallet.currencyInfo.currencyCode` is `ETH`

Reply:
Look please. I added checkEthOnly function to check ETH tokens